// appium-youiengine-driver/build/lib/ios-deploy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _teen_process = require("teen_process");

var _appiumSupport = require("appium-support");

var _logger = _interopRequireDefault(require("./logger"));

var _asyncbox = require("asyncbox");

const IOSDEPLOY_PATH = `ios-deploy`;

class IOSDeploy {
  constructor(udid) {
    this.udid = udid;
    this.cmd = IOSDEPLOY_PATH;
  }

  async checkStatus() {
    await _appiumSupport.fs.which(this.cmd);
  }

  async launchApp(app) {
    let remove = [`--justlaunch`, `--no-wifi`, `--noinstall`, this.udid, `--bundle`, app];

    try {
      await (0, _teen_process.exec)(this.cmd, remove);
    } catch (err) {
      _logger.default.debug(`Stdout: '${err.stdout}'. Stderr: '${err.stderr}'.`);

      throw new Error(`Could not remove app: '${err.message}'`);
    }
  }

  async remove(bundleid) {
    let remove = [`--uninstall_only`, `--no-wifi`, `--id`, this.udid, `--bundle_id`, bundleid];

    try {
      await (0, _teen_process.exec)(this.cmd, remove);
    } catch (err) {
      _logger.default.debug(`Stdout: '${err.stdout}'. Stderr: '${err.stderr}'.`);

      throw new Error(`Could not remove app: '${err.message}'`);
    }
  }

  async removeApp(bundleId) {
    await this.remove(bundleId);
  }

  async install(app) {
    const args = [`--justlaunch`, `--no-wifi`, `--uninstall`, `--id`, this.udid, `--bundle`, app];

    try {
      await (0, _asyncbox.retryInterval)(5, 500, _teen_process.exec, this.cmd, args);
    } catch (err) {
      _logger.default.debug(`Stdout: '${err.stdout}'. Stderr: '${err.stderr}'.`);

      throw new Error(`Could not install app: '${err.message}'`);
    }
  }

  async installApp(app) {
    await this.install(app);
  }

  async isAppInstalled(bundleid) {
    let isInstalled = [`--exists`, `--no-wifi`, `--id`, this.udid, `--bundle_id`, bundleid];

    try {
      let {
        stdout
      } = await (0, _teen_process.exec)(this.cmd, isInstalled);
      return stdout && stdout.includes('true');
    } catch (err) {
      if (err.code !== 255) {
        _logger.default.debug(`Error checking install status: '${err.message}'`);
      }

      return false;
    }
  }

}

var _default = IOSDeploy;
exports.default = _default;require('source-map-support').install();
