// appium-youiengine-driver/build/lib/tvossimulator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _asyncbox = require("asyncbox");

var _appiumIosSimulator = require("appium-ios-simulator");

var simctl = _interopRequireWildcard(require("node-simctl"));

var _basedevice = _interopRequireDefault(require("./basedevice"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

class TvOsSimulator extends _basedevice.default {
  constructor() {
    super();
    this.bundleId;
    this.caps;
    this.shell;
    this.sim;
  }

  execScript(script) {
    _logger.default.debug(`tvOS simulator: execScript`);

    script += ` --id ${this.caps.udid}`;

    try {
      return this.shell.exec(script);
    } catch (err) {
      _logger.default.debug(`Stdout: '${err.stdout}'. Stderr: '${err.stderr}'.`);

      throw new Error(`Could not run '${script}': '${err.message}'`);
    }
  }

  async closeApp() {
    _logger.default.info(`tvOS simulator: Close app`);

    await simctl.terminate(this.caps.udid, this.bundleId);
  }

  async endSession() {
    _logger.default.info(`tvOS simulator: End Session`);

    if (this.caps.fullReset) {
      await this.removeApp(this.bundleId);
    } else {
      await this.closeApp();
    }
  }

  async installApp(appPath) {
    _logger.default.info(`tvOS simulator: Installing and launching app`);

    this.bundleId = this.shell.exec(`osascript -e 'id of app "${appPath}"'`).replace(/(\r\n|\n|\r)/gm, '');
    let retry = false;

    do {
      try {
        await this.sim.installApp(appPath);
        retry = false;
      } catch (err) {
        retry = true;
      }
    } while (retry === true);

    await simctl.launch(this.caps.udid, this.bundleId);
  }

  async isAppInstalled(bundleId) {
    _logger.default.info(`tvOS simulator: Check if App is installed`);

    return await this.sim.isAppInstalled(bundleId);
  }

  async launchApp() {
    _logger.default.info(`tvOS simulator: Launching app`);

    await simctl.launch(this.caps.udid, this.bundleId);
  }

  async removeApp(bundleId) {
    _logger.default.info(`tvOS simulator: Deleting app`);

    await this.sim.removeApp(bundleId);
  }

  async startSession(caps) {
    _logger.default.info(`tvOS simulator: Start Session`);

    this.caps = caps;
    this.shell = require('shelljs');
    this.sim = await (0, _appiumIosSimulator.getSimulator)(this.caps.udid);
    await this.sim.run({
      startupTimeout: 10000
    });
    await (0, _asyncbox.sleep)(4000);
    this.bundleId = this.shell.exec(`osascript -e 'id of app "${caps.app}"'`).replace(/(\r\n|\n|\r)/gm, '');
    let devAppInstalled = await this.isAppInstalled(this.bundleId);

    if (caps.fullReset || !devAppInstalled) {
      await this.installApp(caps.app);
    } else {
      await this.launchApp(caps.app);
    }
  }

}

var _default = TvOsSimulator;
exports.default = _default;require('source-map-support').install();
