// appium-youiengine-driver/build/lib/yilinux.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _basedevice = _interopRequireDefault(require("./basedevice"));

class YiLinux extends _basedevice.default {
  constructor() {
    super();
    this.caps;
    this.shell;
  }

  async closeApp() {
    _logger.default.info(`YiLinux: Close App`);

    let process_name = this.caps.app.substring(this.caps.app.lastIndexOf('/') + 1);
    await this.shell.exec(`killall ${process_name}`);
  }

  async endSession() {
    _logger.default.info(`YiLinux: End Session`);

    await this.closeApp();
  }

  async launchApp() {
    _logger.default.info(`YiLinux: Launch app`);

    let spawn = require('child_process').spawn,
        ls = await spawn(this.caps.app);

    ls.stdout.on('data', function (data) {
      if (data != null) {
        _logger.default.debug(` Log Output: ${data.toString()}`);
      }
    });
    ls.stderr.on('data', function (data) {
      if (data != null) {
        _logger.default.debug(` Log Error: ${data.toString()}`);
      }
    });
    ls.on('exit', function (code) {
      if (code != null) {
        _logger.default.debug(`Application exited with code ${code.toString()}`);
      }
    });
  }

  async startSession(caps) {
    _logger.default.info(`YiLinux: Start Session`);

    this.caps = caps;
    this.shell = require('shelljs');
    await this.closeApp();
    await this.launchApp();
  }

}

var _default = YiLinux;
exports.default = _default;require('source-map-support').install();
