// appium-espresso-driver/build/lib/desired-caps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _appiumAndroidDriver = require("appium-android-driver");

let espressoCapConstraints = {
  app: {
    isString: true
  },
  automationName: {
    isString: true
  },
  systemPort: {
    isNumber: true
  },
  browserName: {
    isString: true
  },
  launchTimeout: {
    isNumber: true
  },
  skipUnlock: {
    isBoolean: true
  },
  forceEspressoRebuild: {
    isBoolean: true
  },
  espressoServerLaunchTimeout: {
    isNumber: true
  },
  espressoBuildConfig: {
    isString: true
  },
  showGradleLog: {
    isBoolean: true
  },
  skipServerInstallation: {
    isBoolean: true
  },
  intentOptions: {
    isObject: true
  },
  disableSuppressAccessibilityService: {
    isBoolean: true
  },
  activityOptions: {
    isObject: true
  },
  appLocale: {
    isObject: true
  }
};
let desiredCapConstraints = {};
Object.assign(desiredCapConstraints, espressoCapConstraints, _appiumAndroidDriver.commonCapConstraints);
var _default = desiredCapConstraints;
exports.default = _default;require('source-map-support').install();
