// appium-idb/build/lib/helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPids = getPids;
exports.fixOutputToArray = fixOutputToArray;
exports.fixOutputToObject = fixOutputToObject;
exports.convertToIDBEnv = convertToIDBEnv;
exports.DEFAULT_COMPANION_GRPC_PORT = exports.DEFAULT_COMPANION_PORT = exports.DEFAULT_IDB_PORT = exports.IDB_COMPANION_EXECUTABLE = exports.IDB_EXECUTABLE = exports.DEFAULT_IDB_EXEC_TIMEOUT = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _teen_process = require("teen_process");

const DEFAULT_IDB_EXEC_TIMEOUT = 20000;
exports.DEFAULT_IDB_EXEC_TIMEOUT = DEFAULT_IDB_EXEC_TIMEOUT;
const DEFAULT_IDB_PORT = 9889;
exports.DEFAULT_IDB_PORT = DEFAULT_IDB_PORT;
const IDB_EXECUTABLE = 'idb';
exports.IDB_EXECUTABLE = IDB_EXECUTABLE;
const IDB_COMPANION_EXECUTABLE = 'idb_companion';
exports.IDB_COMPANION_EXECUTABLE = IDB_COMPANION_EXECUTABLE;
const DEFAULT_COMPANION_PORT = 10880;
exports.DEFAULT_COMPANION_PORT = DEFAULT_COMPANION_PORT;
const DEFAULT_COMPANION_GRPC_PORT = 10882;
exports.DEFAULT_COMPANION_GRPC_PORT = DEFAULT_COMPANION_GRPC_PORT;
const IDB_ENV_PREFIX = 'IDB_';

async function getPids(pattern, opts = {}) {
  const {
    multi = true,
    ignoreCase = true
  } = opts;
  const args = [`-${ignoreCase ? 'i' : ''}f${multi ? '' : 'n'}`, pattern];

  try {
    const {
      stdout
    } = await (0, _teen_process.exec)('pgrep', args);
    const result = stdout.split('\n').filter(Number).map(x => `${x}`);
    return multi ? result : _lodash.default.isEmpty(result) ? [] : _lodash.default.first(result);
  } catch (err) {
    return [];
  }
}

function convertToIDBEnv(env) {
  if (!_lodash.default.isPlainObject(env) || _lodash.default.isEmpty(env)) {
    return null;
  }

  return _lodash.default.reduce(env, (result, value, key) => {
    result[IDB_ENV_PREFIX + key] = value;
    return result;
  }, {});
}

function fixOutputToArray(output) {
  if (!_lodash.default.trim(output)) {
    return [];
  }

  return output.split('\n').reduce((acc, x) => {
    try {
      return [...acc, JSON.parse(x)];
    } catch (e) {
      return acc;
    }
  }, []);
}

function fixOutputToObject(output) {
  if (!_lodash.default.trim(output)) {
    return {};
  }

  const result = {};
  const lines = output.split('\n');

  const getLeftIndent = line => line.length - _lodash.default.trimStart(line).length;

  let lineIdx = 0;

  do {
    if (!_lodash.default.trim(lines[lineIdx])) {
      lineIdx++;
      continue;
    }

    const objectMatch = /(\S+)\s+{/.exec(lines[lineIdx]);

    if (objectMatch) {
      const currentIndent = getLeftIndent(lines[lineIdx]);
      const startLine = lineIdx;

      do {
        lineIdx++;
      } while (lineIdx < lines.length && currentIndent < getLeftIndent(lines[lineIdx]));

      const objectName = objectMatch[1];
      const objectContent = lines.slice(startLine + 1, lineIdx).join('\n');
      result[objectName] = fixOutputToObject(objectContent);
    }

    const propertyMatch = /(\S+):\s+([^\n]+)/.exec(lines[lineIdx]);

    if (propertyMatch) {
      const propertyName = propertyMatch[1];
      const propertyValue = propertyMatch[2].trim();
      result[propertyName] = propertyValue.startsWith('"') ? _lodash.default.trim(propertyValue, '"') : Number(propertyValue);
    }

    lineIdx++;
  } while (lineIdx < lines.length);

  return result;
}require('source-map-support').install();
