// appium-safari-driver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _logger = _interopRequireDefault(require("./logger"));

var _safari = _interopRequireDefault(require("./safari"));

var _desiredCaps = require("./desired-caps");

var _index = _interopRequireDefault(require("./commands/index"));

var _utils = require("./utils");

const NO_PROXY = [['GET', new RegExp('^/session/[^/]+/appium')], ['POST', new RegExp('^/session/[^/]+/appium')], ['POST', new RegExp('^/session/[^/]+/element/[^/]+/elements?$')], ['POST', new RegExp('^/session/[^/]+/elements?$')]];

class SafariDriver extends _appiumBaseDriver.BaseDriver {
  constructor(opts = {}) {
    super(opts);
    this.desiredCapConstraints = _desiredCaps.desiredCapConstraints;
    this.locatorStrategies = ['xpath', 'tag name', 'link text', 'partial link text', 'css selector', 'id', 'name'];
    this.resetState();

    for (const [cmd, fn] of _lodash.default.toPairs(_index.default)) {
      SafariDriver.prototype[cmd] = fn;
    }
  }

  resetState() {
    this.safari = null;
    this.proxyReqRes = null;
    this.isProxyActive = false;
    this._screenRecorder = null;
  }

  proxyActive() {
    return this.isProxyActive;
  }

  getProxyAvoidList() {
    return NO_PROXY;
  }

  canProxy() {
    return true;
  }

  async createSession(...args) {
    const [sessionId, caps] = await super.createSession(...args);
    this.safari = new _safari.default();

    try {
      await this.safari.start((0, _utils.formatCapsForServer)(caps));
    } catch (e) {
      await this.deleteSession();
      throw e;
    }

    this.proxyReqRes = this.safari.proxy.proxyReqRes.bind(this.safari.proxy);
    this.isProxyActive = true;
    return [sessionId, caps];
  }

  async deleteSession() {
    var _this$_screenRecorder, _this$safari;

    _logger.default.info('Ending Safari session');

    await ((_this$_screenRecorder = this._screenRecorder) === null || _this$_screenRecorder === void 0 ? void 0 : _this$_screenRecorder.stop(true));
    await ((_this$safari = this.safari) === null || _this$safari === void 0 ? void 0 : _this$safari.stop());
    this.resetState();
    await super.deleteSession();
  }

}

var _default = SafariDriver;
exports.default = _default;require('source-map-support').install();
