// appium-windows-driver/build/lib/commands/execute.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _logger = _interopRequireDefault(require("../logger"));

var _constants = require("../constants");

const commands = {};
const POWER_SHELL_SCRIPT_PATTERN = /^powerShell$/;
const WINDOWS_EXTENSION_SCRIPT_PATTERN = /^windows:/;
const EXTENSION_COMMANDS_MAPPING = {
  startRecordingScreen: 'startRecordingScreen',
  stopRecordingScreen: 'stopRecordingScreen',
  deleteFile: 'windowsDeleteFile',
  deleteFolder: 'windowsDeleteFolder'
};

commands.execute = async function execute(script, args) {
  if (WINDOWS_EXTENSION_SCRIPT_PATTERN.test(script)) {
    _logger.default.info(`Executing extension command '${script}'`);

    script = script.replace(WINDOWS_EXTENSION_SCRIPT_PATTERN, '').trim();
    return await this.executeWindowsCommand(script, _lodash.default.isArray(args) ? args[0] : args);
  } else if (POWER_SHELL_SCRIPT_PATTERN.test(script)) {
    this.ensureFeatureEnabled(_constants.POWER_SHELL_FEATURE);
    return await this.execPowerShell(_lodash.default.isArray(args) ? _lodash.default.first(args) : args);
  }

  throw new _appiumBaseDriver.errors.NotImplementedError();
};

commands.executeWindowsCommand = async function executeWindowsCommand(command, opts = {}) {
  if (!_lodash.default.has(EXTENSION_COMMANDS_MAPPING, command)) {
    throw new _appiumBaseDriver.errors.UnknownCommandError(`Unknown windows command '${command}'. ` + `Only ${_lodash.default.keys(EXTENSION_COMMANDS_MAPPING)} commands are supported.`);
  }

  return await this[EXTENSION_COMMANDS_MAPPING[command]](opts);
};

var _default = commands;
exports.default = _default;require('source-map-support').install();
