// appium-windows-driver/build/lib/commands/file-movement.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.commands = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _logger = _interopRequireDefault(require("../logger"));

var _path = _interopRequireDefault(require("path"));

var _appiumBaseDriver = require("appium-base-driver");

var _appiumSupport = require("appium-support");

var _constants = require("../constants");

const KNOWN_ENV_VARS = ['APPDATA', 'LOCALAPPDATA', 'PROGRAMFILES', 'PROGRAMFILES(X86)', 'PROGRAMDATA', 'ALLUSERSPROFILE', 'TEMP', 'TMP', 'HOMEPATH', 'USERPROFILE', 'PUBLIC'];
const commands = {};
exports.commands = commands;

commands.pushFile = async function pushFile(remotePath, base64Data) {
  this.ensureFeatureEnabled(_constants.MODIFY_FS_FEATURE);

  if (remotePath.endsWith(_path.default.sep)) {
    throw new _appiumBaseDriver.errors.InvalidArgumentError('It is expected that remote path points to a file rather than a folder. ' + `'${remotePath}' is given instead`);
  }

  if (_lodash.default.isArray(base64Data)) {
    base64Data = Buffer.from(base64Data).toString('utf8');
  }

  const fullPath = resolveToAbsolutePath(remotePath);
  await (0, _appiumSupport.mkdirp)(_path.default.dirname(fullPath));
  const content = Buffer.from(base64Data, 'base64');
  await _appiumSupport.fs.writeFile(fullPath, content);
};

commands.pullFile = async function pullFile(remotePath) {
  const fullPath = resolveToAbsolutePath(remotePath);
  await checkFileExists(fullPath);
  return (await _appiumSupport.util.toInMemoryBase64(fullPath)).toString();
};

commands.pullFolder = async function pullFolder(remotePath) {
  const fullPath = resolveToAbsolutePath(remotePath);
  await checkFolderExists(fullPath);
  return (await _appiumSupport.zip.toInMemoryZip(fullPath, {
    encodeToBase64: true
  })).toString();
};

commands.windowsDeleteFile = async function windowsDeleteFile(opts = {}) {
  this.ensureFeatureEnabled(_constants.MODIFY_FS_FEATURE);
  const {
    remotePath
  } = opts;
  const fullPath = resolveToAbsolutePath(remotePath);
  await checkFileExists(fullPath);
  await _appiumSupport.fs.unlink(fullPath);
};

commands.windowsDeleteFolder = async function windowsDeleteFolder(opts = {}) {
  this.ensureFeatureEnabled(_constants.MODIFY_FS_FEATURE);
  const {
    remotePath
  } = opts;
  const fullPath = resolveToAbsolutePath(remotePath);
  await checkFolderExists(fullPath);
  await _appiumSupport.fs.rimraf(fullPath);
};

function resolveToAbsolutePath(remotePath) {
  const resolvedPath = remotePath.replace(/%([^%]+)%/g, (_, key) => KNOWN_ENV_VARS.includes(key.toUpperCase()) ? process.env[key.toUpperCase()] : `%${key}%`);

  _logger.default.debug(`Resolved path '${resolvedPath}'.`);

  if (!_path.default.isAbsolute(resolvedPath)) {
    throw new _appiumBaseDriver.errors.InvalidArgumentError('It is expected that remote path is absolute. ' + `'${resolvedPath}' is given instead`);
  }

  return resolvedPath;
}

async function checkFileExists(remotePath) {
  if (!(await _appiumSupport.fs.exists(remotePath))) {
    throw new _appiumBaseDriver.errors.InvalidArgumentError(`The remote file '${remotePath}' does not exist.`);
  }

  const stat = await _appiumSupport.fs.stat(remotePath);

  if (!stat.isFile()) {
    throw new _appiumBaseDriver.errors.InvalidArgumentError('It is expected that remote path points to a file rather than a folder. ' + `'${remotePath}' is given instead`);
  }
}

async function checkFolderExists(remotePath) {
  if (!(await _appiumSupport.fs.exists(remotePath))) {
    throw new _appiumBaseDriver.errors.InvalidArgumentError(`The remote folder '${remotePath}' does not exist.`);
  }

  const stat = await _appiumSupport.fs.stat(remotePath);

  if (!stat.isDirectory()) {
    throw new _appiumBaseDriver.errors.InvalidArgumentError('It is expected that remote path points to a folder rather than a file. ' + `'${remotePath}' is given instead`);
  }
}

var _default = commands;
exports.default = _default;require('source-map-support').install();
