// appium-ios-driver/build/lib/instruments/instruments.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _teen_process = require("teen_process");

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

var _through = require("through");

var _path = _interopRequireDefault(require("path"));

var _appiumSupport = require("appium-support");

var _appiumXcode = _interopRequireDefault(require("appium-xcode"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _appiumIosSimulator = require("appium-ios-simulator");

var _utils = require("./utils");

var _streams = require("./streams");

require("colors");

const ERR_NEVER_CHECKED_IN = 'Instruments never checked in';
const ERR_CRASHED_ON_STARTUP = 'Instruments crashed on startup';
const ERR_AMBIGUOUS_DEVICE = 'Instruments Usage Error : Ambiguous device name/identifier';

class Instruments {
  static async quickInstruments(opts) {
    opts = _lodash.default.clone(opts);
    let xcodeTraceTemplatePath = await _appiumXcode.default.getAutomationTraceTemplatePath();

    _lodash.default.defaults(opts, {
      launchTimeout: 60000,
      template: xcodeTraceTemplatePath,
      withoutDelay: true,
      xcodeVersion: '8.1',
      webSocket: null,
      flakeyRetries: 2
    });

    return new Instruments(opts);
  }

  constructor(opts) {
    opts = _lodash.default.cloneDeep(opts);

    _lodash.default.defaults(opts, {
      termTimeout: 5000,
      tmpDir: '/tmp/appium-instruments',
      launchTimeout: 90000,
      flakeyRetries: 0,
      realDevice: false
    });

    const props = ['app', 'termTimeout', 'flakeyRetries', 'udid', 'bootstrap', 'template', 'withoutDelay', 'processArguments', 'realDevice', 'simulatorSdkAndDevice', 'tmpDir', 'traceDir', 'locale', 'language'];

    for (const f of props) {
      this[f] = opts[f];
    }

    this.traceDir = this.traceDir || this.tmpDir;
    this.launchTimeout = (0, _utils.parseLaunchTimeout)(opts.launchTimeout);
    this.proc = null;
    this.webSocket = opts.webSocket;
    this.instrumentsPath = opts.instrumentsPath;
    this.launchTries = 0;
    this.socketConnectDelays = [];
    this.gotFBSOpenApplicationError = false;
    this.onShutdown = new _bluebird.default((resolve, reject) => {
      this.onShutdownDeferred = {
        resolve,
        reject
      };
    });
    this.onShutdown.catch(() => {}).done();
  }

  async configure() {
    if (!this.xcodeVersion) {
      this.xcodeVersion = await _appiumXcode.default.getVersion(true);
    }

    if (this.xcodeVersion.major < 5 && this.withoutDelay) {
      _logger.default.info('instruments-without-delay for Xcode 4 has been removed ' + 'because of macOS notarization. ' + 'You can disable instruments-without-delay ' + 'with the --native-instruments-lib server flag');
    }

    if (this.xcodeVersion.versionFloat === 6.0 && this.withoutDelay) {
      _logger.default.info('In xcode 6.0, instruments-without-delay does not work. ' + 'If using Appium, you can disable instruments-without-delay ' + 'with the --native-instruments-lib server flag');
    }

    if (this.xcodeVersion.versionString === '5.0.1') {
      throw new Error('Xcode 5.0.1 ships with a broken version of ' + 'Instruments. please upgrade to 5.0.2');
    }

    if (this.xcodeVersion.major > 7) {
      throw new Error(`Instruments-based automation was removed in Xcode 8. ` + `Xcode ${this.xcodeVersion.versionString} is not supported. ` + `Please try the XCUItest driver.`);
    }

    if (!this.template) {
      this.template = await _appiumXcode.default.getAutomationTraceTemplatePath();
    }

    if (!this.instrumentsPath) {
      this.instrumentsPath = await (0, _utils.getInstrumentsPath)();
    }
  }

  async launchOnce() {
    _logger.default.info('Launching instruments');

    await _appiumSupport.fs.rimraf(this.tmpDir);
    await (0, _appiumSupport.mkdirp)(this.tmpDir);
    await (0, _appiumSupport.mkdirp)(this.traceDir);
    this.exitListener = null;
    this.proc = await this.spawnInstruments();
    this.proc.on('exit', (code, signal) => {
      const msg = code !== null ? `code: ${code}` : `signal: ${signal}`;

      _logger.default.debug(`Instruments exited with ${msg}`);
    });
    let launchResultPromise = new _bluebird.default((resolve, reject) => {
      this.launchResultDeferred = {
        resolve,
        reject
      };
    });
    this.setExitListener(() => {
      this.proc = null;
      this.launchResultDeferred.reject(new Error(ERR_CRASHED_ON_STARTUP));
    });
    this.proc.on('error', err => {
      _logger.default.debug(`Error with instruments proc: ${err.message}`);

      if (err.message.indexOf('ENOENT') !== -1) {
        this.proc = null;

        _logger.default.error(`Unable to spawn instruments: ${err.message}`);

        this.launchResultDeferred.reject(err);
      }
    });
    this.proc.stdout.setEncoding('utf8');
    this.proc.stdout.pipe((0, _streams.outputStream)()).pipe((0, _streams.dumpStream)());
    this.proc.stderr.setEncoding('utf8');

    let actOnStderr = output => {
      if (this.launchTimeout.afterSimLaunch && output && output.match(/CLTilesManagerClient: initialize/)) {
        this.addSocketConnectTimer(this.launchTimeout.afterSimLaunch, 'afterLaunch', async () => {
          await this.killInstruments();
          this.launchResultDeferred.reject(new Error(ERR_NEVER_CHECKED_IN));
        });
      }

      let fbsErrStr = '(FBSOpenApplicationErrorDomain error 8.)';

      if (output.indexOf(fbsErrStr) !== -1) {
        this.gotFBSOpenApplicationError = true;
      }

      if (output.indexOf(ERR_AMBIGUOUS_DEVICE) !== -1) {
        let msg = `${ERR_AMBIGUOUS_DEVICE}: '${this.simulatorSdkAndDevice}'`;
        this.launchResultDeferred.reject(new Error(msg));
      }
    };

    this.proc.stderr.pipe((0, _through.through)(function onStderr(output) {
      actOnStderr(output);
      this.queue(output);
    })).pipe((0, _streams.errorStream)()).pipe((0, _streams.webSocketAlertStream)(this.webSocket)).pipe((0, _streams.dumpStream)());
    this.addSocketConnectTimer(this.launchTimeout.global, 'global', async () => {
      await this.killInstruments();
      this.launchResultDeferred.reject(new Error(ERR_NEVER_CHECKED_IN));
    });

    try {
      await launchResultPromise;
    } finally {
      this.clearSocketConnectTimers();
    }

    this.setExitListener((code, signal) => {
      this.proc = null;
      const msg = code !== null ? `code: ${code}` : `signal: ${signal}`;
      this.onShutdownDeferred.reject(new Error(`Abnormal exit with ${msg}`));
    });
  }

  async launch() {
    await this.configure();
    let launchTries = 0;

    do {
      launchTries++;

      _logger.default.debug(`Attempting to launch instruments, this is try #${launchTries}`);

      try {
        await this.launchOnce();
        break;
      } catch (err) {
        _logger.default.error(`Error launching instruments: ${err.message}`);

        let errIsCatchable = err.message === ERR_NEVER_CHECKED_IN || err.message === ERR_CRASHED_ON_STARTUP;

        if (!errIsCatchable) {
          throw err;
        }

        if (launchTries <= this.flakeyRetries) {
          if (this.gotFBSOpenApplicationError) {
            _logger.default.debug('Got the FBSOpenApplicationError, not killing the ' + 'sim but leaving it open so the app will launch');

            this.gotFBSOpenApplicationError = false;
            await _bluebird.default.delay(1000);
          } else {
            if (!this.realDevice) {
              await (0, _appiumIosSimulator.killAllSimulators)();
            }

            await _bluebird.default.delay(5000);
          }
        } else {
          _logger.default.errorAndThrow('We exceeded the number of retries allowed for ' + 'instruments to successfully start; failing launch');
        }
      }
    } while (true);
  }

  registerLaunch() {
    this.launchResultDeferred.resolve();
  }

  async spawnInstruments() {
    let traceDir;

    for (let i = 0;; i++) {
      traceDir = _path.default.resolve(this.traceDir, `instrumentscli${i}.trace`);
      if (!(await _appiumSupport.fs.exists(traceDir))) break;
    }

    let args = ['-t', this.template, '-D', traceDir];

    if (this.udid) {
      args = args.concat(['-w', this.udid]);

      _logger.default.debug(`Attempting to run app on real device with UDID '${this.udid}'`);
    }

    if (!this.udid && this.simulatorSdkAndDevice) {
      args = args.concat(['-w', this.simulatorSdkAndDevice]);

      _logger.default.debug(`Attempting to run app on ${this.simulatorSdkAndDevice}`);
    }

    args = args.concat([this.app]);

    if (this.processArguments) {
      _logger.default.debug(`Attempting to run app with process arguments: ${JSON.stringify(this.processArguments)}`);

      if (_lodash.default.isString(this.processArguments)) {
        if (this.processArguments.indexOf('-e ') === -1) {
          _logger.default.debug('Plain string process arguments being pushed into arguments');

          args.push(this.processArguments);
        } else {
          _logger.default.debug('Environment variables being pushed into arguments');

          for (let arg of this.processArguments.split('-e ')) {
            arg = arg.trim();

            if (arg.length) {
              let space = arg.indexOf(' ');
              let flag = arg.substring(0, space);
              let value = arg.substring(space + 1);
              args.push('-e', flag, value);
            }
          }
        }
      } else {
        for (let [flag, value] of _lodash.default.toPairs(this.processArguments)) {
          args.push('-e', flag, value);
        }
      }
    }

    args = args.concat(['-e', 'UIASCRIPT', this.bootstrap]);
    args = args.concat(['-e', 'UIARESULTSPATH', this.tmpDir]);

    if (this.language) {
      args = args.concat([`-AppleLanguages (${this.language})`]);
      args = args.concat([`-NSLanguages (${this.language})`]);
    }

    if (this.locale) {
      args = args.concat([`-AppleLocale ${this.locale}`]);
    }

    let env = _lodash.default.clone(process.env);

    if (this.xcodeVersion.major >= 7 && !this.udid) {
      _logger.default.info('On xcode 7.0+, instruments-without-delay does not work, ' + 'skipping instruments-without-delay');

      this.withoutDelay = false;
    }

    let iwdPath = await (0, _utils.getIwdPath)(this.xcodeVersion.major);
    env.CA_DEBUG_TRANSACTIONS = 1;

    if (this.withoutDelay && !this.udid) {
      env.DYLD_INSERT_LIBRARIES = _path.default.resolve(iwdPath, 'InstrumentsShim.dylib');
      env.LIB_PATH = iwdPath;
    }

    let instrumentsExecArgs = [this.instrumentsPath, ...args];
    instrumentsExecArgs = _lodash.default.map(instrumentsExecArgs, function instrumentArgs(arg) {
      if (arg === null) {
        throw new Error('A null value was passed as an arg to execute ' + 'instruments on the command line. A letiable is ' + 'probably not getting set. Array of command args: ' + JSON.stringify(instrumentsExecArgs));
      }

      if (_lodash.default.isString(arg) && arg.indexOf(' ') !== -1) {
        return `"${arg}"`;
      }

      return arg;
    });

    _logger.default.debug(`Spawning instruments with command: '${instrumentsExecArgs.join(' ')}'`);

    if (this.withoutDelay) {
      _logger.default.debug('And extra without-delay env: ' + JSON.stringify({
        DYLD_INSERT_LIBRARIES: env.DYLD_INSERT_LIBRARIES,
        LIB_PATH: env.LIB_PATH
      }));
    }

    _logger.default.debug(`And launch timeouts (in ms): ${JSON.stringify(this.launchTimeout)}`);

    return await (0, _teen_process.spawn)(this.instrumentsPath, args, {
      env
    });
  }

  addSocketConnectTimer(delay, type, doAction) {
    let socketConnectDelay = (0, _appiumSupport.cancellableDelay)(delay);
    socketConnectDelay.then(() => {
      _logger.default.warn(`Instruments socket client never checked in; timing out (${type})`);

      return doAction();
    }).catch(_bluebird.default.CancellationError, () => {}).done();
    this.socketConnectDelays.push(socketConnectDelay);
  }

  clearSocketConnectTimers() {
    for (let delay of this.socketConnectDelays) {
      delay.cancel();
    }

    this.socketConnectDelays = [];
  }

  setExitListener(exitListener) {
    if (!this.proc) return;

    if (this.exitListener) {
      this.proc.removeListener('exit', this.exitListener);
    }

    this.exitListener = exitListener;
    this.proc.on('exit', exitListener);
  }

  killInstruments() {
    if (!this.proc) return;

    _logger.default.debug(`Kill Instruments process (pid: ${this.proc.pid})`);

    return new _bluebird.default(async resolve => {
      let wasTerminated = false;
      let termDelay = (0, _appiumSupport.cancellableDelay)(this.termTimeout);
      let termPromise = termDelay.catch(_bluebird.default.CancellationError, () => {});
      this.setExitListener(() => {
        this.proc = null;
        wasTerminated = true;
        termDelay.cancel();
        resolve();
      });

      _logger.default.debug('Sending SIGTERM');

      this.proc.kill('SIGTERM');
      await termPromise;

      if (!wasTerminated) {
        _logger.default.warn(`Instruments did not terminate after ${this.termTimeout / 1000} seconds!`);

        _logger.default.debug('Sending SIGKILL');

        this.proc.kill('SIGKILL');

        if (_lodash.default.isFunction(this.exitListener)) {
          this.exitListener();
        }
      }
    });
  }

  async shutdown() {
    _logger.default.debug('Starting shutdown.');

    await this.killInstruments();
    this.onShutdownDeferred.resolve();
  }

}

var _default = Instruments;
exports.default = _default;require('source-map-support').install();
