// appium-xcuitest-driver/build/lib/cert-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.doesSupportKeychainApi = void 0;
exports.hasCertificateLegacy = hasCertificateLegacy;
exports.installCertificate = installCertificate;
exports.installCertificateLegacy = installCertificateLegacy;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumIosSimulator = require("appium-ios-simulator");

const doesSupportKeychainApi = _lodash.default.memoize(async function doesSupportKeychainApi(device) {
  try {
    await device.simctl.exec('help', {
      args: ['keychain']
    });
    return true;
  } catch (e) {
    return false;
  }
});

exports.doesSupportKeychainApi = doesSupportKeychainApi;

async function installCertificate(device, payload) {
  await device.simctl.addRootCertificate(payload, {
    raw: true
  });
}

async function hasCertificateLegacy(device, payload) {
  return await (0, _appiumIosSimulator.hasSSLCert)(payload, device.udid);
}

async function installCertificateLegacy(device, payload) {
  await (0, _appiumIosSimulator.installSSLCert)(payload, device.udid);
}require('source-map-support').install();
