// appium-ios-device/build/lib/usbmux/transformer/usbmux-decoder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UsbmuxDecoder = void 0;

require("source-map-support/register");

var _stream = _interopRequireDefault(require("stream"));

var _appiumSupport = require("appium-support");

const HEADER_LENGTH = 16;

class UsbmuxDecoder extends _stream.default.Transform {
  constructor() {
    super({
      objectMode: true
    });
  }

  _transform(data, encoding, callback) {
    this._decode(data);

    callback();
  }

  _decode(data) {
    const header = {
      length: data.readUInt32LE(0),
      version: data.readUInt32LE(4),
      type: data.readUInt32LE(8),
      tag: data.readUInt32LE(12)
    };
    let payload = data.slice(HEADER_LENGTH, data.length);
    this.push({
      header,
      payload: _appiumSupport.plist.parsePlist(payload)
    });
  }

}

exports.UsbmuxDecoder = UsbmuxDecoder;
var _default = UsbmuxDecoder;
exports.default = _default;require('source-map-support').install();
