// appium-mac2-driver/build/lib/commands/execute.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _logger = _interopRequireDefault(require("../logger"));

const commands = {};
const EXTENSION_COMMANDS_MAPPING = {
  setValue: 'macosSetValue',
  click: 'macosClick',
  scroll: 'macosScroll',
  swipe: 'macosSwipe',
  rightClick: 'macosRightClick',
  hover: 'macosHover',
  doubleClick: 'macosDoubleClick',
  clickAndDrag: 'macosClickAndDrag',
  clickAndDragAndHold: 'macosClickAndDragAndHold',
  keys: 'macosKeys',
  tap: 'macosTap',
  doubleTap: 'macosDoubleTap',
  press: 'macosPress',
  pressAndDrag: 'macosPressAndDrag',
  pressAndDragAndHold: 'macosPressAndDragAndHold',
  source: 'macosSource',
  launchApp: 'macosLaunchApp',
  activateApp: 'macosActivateApp',
  terminateApp: 'macosTerminateApp',
  queryAppState: 'macosQueryAppState',
  appleScript: 'macosExecAppleScript',
  startRecordingScreen: 'startRecordingScreen',
  stopRecordingScreen: 'stopRecordingScreen'
};

commands.execute = async function execute(script, args) {
  if (script.match(/^macos:/)) {
    _logger.default.info(`Executing extension command '${script}'`);

    script = script.replace(/^macos:/, '').trim();
    return await this.executeMacosCommand(script, _lodash.default.isArray(args) ? args[0] : args);
  }

  throw new _appiumBaseDriver.errors.NotImplementedError();
};

commands.executeMacosCommand = async function executeMacosCommand(command, opts = {}) {
  if (!_lodash.default.has(EXTENSION_COMMANDS_MAPPING, command)) {
    throw new _appiumBaseDriver.errors.UnknownCommandError(`Unknown extension command "${command}". ` + `Only ${_lodash.default.keys(EXTENSION_COMMANDS_MAPPING)} commands are supported.`);
  }

  return await this[EXTENSION_COMMANDS_MAPPING[command]](opts);
};

var _default = commands;
exports.default = _default;require('source-map-support').install();
