// appium-sdb/build/lib/sdb.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_SDB_PORT = exports.default = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _index = _interopRequireDefault(require("./tools/index.js"));

const DEFAULT_SDB_PORT = 26099;
exports.DEFAULT_SDB_PORT = DEFAULT_SDB_PORT;
const DEFAULT_OPTS = {
  sdkRoot: null,
  udid: null,
  executable: {
    path: 'sdb',
    defaultArgs: []
  },
  curDeviceId: null,
  emulatorPort: null,
  binaries: {},
  suppressKillServer: null,
  sdbPort: DEFAULT_SDB_PORT
};

class SDB {
  constructor(opts = {}) {
    if (typeof opts.sdkRoot === 'undefined') {
      opts.sdkRoot = process.env.TIZEN_HOME || '';
    }

    Object.assign(this, opts);

    _lodash.default.defaultsDeep(this, _lodash.default.cloneDeep(DEFAULT_OPTS));

    if (opts.remoteSdbPort) {
      this.sdbPort = opts.remoteSdbPort;
    }
  }

}

SDB.createSDB = async function (opts) {
  let sdb = new SDB(opts);
  await sdb.getSdbWithCorrectSdbPath();
  return sdb;
};

for (let [fnName, fn] of _lodash.default.toPairs(_index.default)) {
  SDB.prototype[fnName] = fn;
}

var _default = SDB;
exports.default = _default;require('source-map-support').install();
