// appium-ios-driver/build/index.js
#!/usr/bin/env node
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "authorize", {
  enumerable: true,
  get: function () {
    return _authorizeIos.authorize;
  }
});
exports.appUtils = exports.uiauto = exports.utils = exports.device = exports.settings = exports.IOSPerformanceLog = exports.IOSCrashLog = exports.IOSLog = exports.default = exports.WEBVIEW_WIN = exports.NATIVE_WIN = exports.startHttpsServer = exports.instrumentsUtils = exports.Instruments = exports.IWDP = exports.defaultServerCaps = exports.iosCommands = exports.commands = exports.desiredCapValidation = exports.desiredCapConstraints = exports.IosDriver = void 0;

require("source-map-support/register");

var _asyncbox = require("asyncbox");

var server = _interopRequireWildcard(require("./lib/server"));

var driver = _interopRequireWildcard(require("./lib/driver"));

var caps = _interopRequireWildcard(require("./lib/desired-caps"));

var commandIndex = _interopRequireWildcard(require("./lib/commands/index"));

var settings = _interopRequireWildcard(require("./lib/settings"));

exports.settings = settings;

var device = _interopRequireWildcard(require("./lib/device"));

exports.device = device;

var utils = _interopRequireWildcard(require("./lib/utils"));

exports.utils = utils;

var iwdp = _interopRequireWildcard(require("./lib/iwdp"));

var uiauto = _interopRequireWildcard(require("./lib/uiauto/uiauto"));

exports.uiauto = uiauto;

var instruments = _interopRequireWildcard(require("./lib/instruments/index"));

var context = _interopRequireWildcard(require("./lib/commands/context"));

var _iosLog = _interopRequireDefault(require("./lib/device-log/ios-log"));

var _iosCrashLog = _interopRequireDefault(require("./lib/device-log/ios-crash-log"));

var _iosPerformanceLog = _interopRequireDefault(require("./lib/device-log/ios-performance-log"));

var appUtils = _interopRequireWildcard(require("./lib/app-utils"));

exports.appUtils = appUtils;

var _authorizeIos = require("./bin/authorize-ios");

const DEFAULT_HOST = 'localhost';
const DEFAULT_PORT = 4723;

async function main() {
  const getArgValue = argName => {
    const argIndex = process.argv.indexOf(argName);
    return argIndex > 0 ? process.argv[argIndex + 1] : null;
  };

  const port = parseInt(getArgValue('--port'), 10) || DEFAULT_PORT;
  const host = getArgValue('--host') || DEFAULT_HOST;
  return await server.startServer(port, host);
}

if (require.main === module) {
  (0, _asyncbox.asyncify)(main);
}

const startHttpsServer = server.startHttpsServer;
exports.startHttpsServer = startHttpsServer;
const {
  IosDriver,
  defaultServerCaps
} = driver;
exports.defaultServerCaps = defaultServerCaps;
exports.IosDriver = IosDriver;
const {
  desiredCapConstraints,
  desiredCapValidation
} = caps;
exports.desiredCapValidation = desiredCapValidation;
exports.desiredCapConstraints = desiredCapConstraints;
const {
  commands,
  iosCommands
} = commandIndex;
exports.iosCommands = iosCommands;
exports.commands = commands;
const {
  IWDP
} = iwdp;
exports.IWDP = IWDP;
const {
  Instruments,
  instrumentsUtils
} = instruments;
exports.instrumentsUtils = instrumentsUtils;
exports.Instruments = Instruments;
const {
  NATIVE_WIN,
  WEBVIEW_WIN
} = context;
exports.WEBVIEW_WIN = WEBVIEW_WIN;
exports.NATIVE_WIN = NATIVE_WIN;
var _default = IosDriver;
exports.default = _default;
const IOSLog = _iosLog.default;
exports.IOSLog = IOSLog;
const IOSCrashLog = _iosCrashLog.default;
exports.IOSCrashLog = IOSCrashLog;
const IOSPerformanceLog = _iosPerformanceLog.default;
exports.IOSPerformanceLog = IOSPerformanceLog;require('source-map-support').install();
