// appium-fake-driver/build/lib/commands/alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _appiumBaseDriver = require("appium-base-driver");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

helpers.assertNoAlert = function assertNoAlert() {
  if (this.appModel.hasAlert()) {
    throw new _appiumBaseDriver.errors.UnexpectedAlertOpenError();
  }
};

helpers.assertAlert = function assertAlert() {
  if (!this.appModel.hasAlert()) {
    throw new _appiumBaseDriver.errors.NoAlertOpenError();
  }
};

commands.getAlertText = async function getAlertText() {
  this.assertAlert();
  return this.appModel.alertText();
};

commands.setAlertText = async function setAlertText(text) {
  this.assertAlert();

  try {
    this.appModel.setAlertText(text);
  } catch (e) {
    throw new _appiumBaseDriver.errors.InvalidElementStateError();
  }
};

commands.postAcceptAlert = async function postAcceptAlert() {
  this.assertAlert();
  this.appModel.handleAlert();
};

commands.postDismissAlert = commands.postAcceptAlert;
Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
