// appium-fake-driver/build/lib/commands/contexts.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

helpers.getRawContexts = function getRawContexts() {
  let contexts = {
    'NATIVE_APP': null
  };
  let wvs = this.appModel.getWebviews();

  for (let i = 1; i < wvs.length + 1; i++) {
    contexts[`WEBVIEW_${i}`] = wvs[i - 1];
  }

  return contexts;
};

helpers.assertWebviewContext = function assertWebviewContext() {
  if (this.curContext === 'NATIVE_APP') {
    throw new _appiumBaseDriver.errors.InvalidContextError();
  }
};

commands.getCurrentContext = async function getCurrentContext() {
  return this.curContext;
};

commands.getContexts = async function getContexts() {
  return _lodash.default.keys(this.getRawContexts());
};

commands.setContext = async function setContext(context) {
  let contexts = this.getRawContexts();

  if (_lodash.default.includes(_lodash.default.keys(contexts), context)) {
    this.curContext = context;

    if (context === 'NATIVE_APP') {
      this.appModel.deactivateWebview();
    } else {
      this.appModel.activateWebview(contexts[context]);
    }
  } else {
    throw new _appiumBaseDriver.errors.NoSuchContextError();
  }
};

commands.setFrame = async function setFrame(frameId) {
  this.assertWebviewContext();

  if (frameId === null) {
    this.appModel.deactivateFrame();
  } else {
    let nodes = this.appModel.xpathQuery(`//iframe[@id="${frameId}"]`);

    if (!nodes.length) {
      throw new _appiumBaseDriver.errors.NoSuchFrameError();
    }

    this.appModel.activateFrame(nodes[0]);
  }
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
