// appium-fake-driver/build/lib/commands/find.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _fakeElement = require("../fake-element");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

helpers.getExistingElementForNode = function getExistingElementForNode(node) {
  for (let [id, el] of _lodash.default.toPairs(this.elMap)) {
    if (el.node === node) {
      return id;
    }
  }

  return null;
};

helpers.wrapNewEl = function wrapNewEl(obj) {
  let existingElId = this.getExistingElementForNode(obj);

  if (existingElId) {
    return {
      ELEMENT: existingElId
    };
  }

  this.maxElId++;
  this.elMap[this.maxElId.toString()] = new _fakeElement.FakeElement(obj, this.appModel);
  return {
    ELEMENT: this.maxElId.toString()
  };
};

helpers.findElOrEls = async function findElOrEls(strategy, selector, mult, ctx) {
  let qMap = {
    'xpath': 'xpathQuery',
    'id': 'idQuery',
    'accessibility id': 'idQuery',
    'class name': 'classQuery',
    'tag name': 'classQuery'
  };

  if (!_lodash.default.includes(_lodash.default.keys(qMap), strategy)) {
    throw new _appiumBaseDriver.errors.UnknownCommandError();
  }

  if (selector === 'badsel') {
    throw new _appiumBaseDriver.errors.InvalidSelectorError();
  }

  let els = this.appModel[qMap[strategy]](selector, ctx);

  if (els.length) {
    if (mult) {
      let allEls = [];

      for (let el of els) {
        allEls.push(this.wrapNewEl(el));
      }

      return allEls;
    } else {
      return this.wrapNewEl(els[0]);
    }
  } else if (mult) {
    return [];
  } else {
    throw new _appiumBaseDriver.errors.NoSuchElementError();
  }
};

commands.findElement = async function findElement(strategy, selector) {
  return this.findElOrEls(strategy, selector, false);
};

commands.findElements = async function findElements(strategy, selector) {
  return this.findElOrEls(strategy, selector, true);
};

commands.findElementFromElement = async function findElementFromElement(strategy, selector, elementId) {
  let el = this.getElement(elementId);
  return this.findElOrEls(strategy, selector, false, el.xmlFragment);
};

commands.findElementsFromElement = async function findElementsFromElement(strategy, selector, elementId) {
  let el = this.getElement(elementId);
  return this.findElOrEls(strategy, selector, true, el.xmlFragment);
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
