// appium-fake-driver/build/lib/commands/general.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

commands.title = async function title() {
  this.assertWebviewContext();
  return this.appModel.title;
};

commands.keys = async function keys(value) {
  if (!this.focusedElId) {
    throw new _appiumBaseDriver.errors.InvalidElementStateError();
  }

  await this.setValue(value, this.focusedElId);
};

commands.setGeoLocation = async function setGeoLocation(location) {
  this.appModel.lat = location.latitude;
  this.appModel.long = location.longitude;
};

commands.getGeoLocation = async function getGeoLocation() {
  return this.appModel.currentGeoLocation;
};

commands.getPageSource = async function getPageSource() {
  return this.appModel.rawXml;
};

commands.getOrientation = async function getOrientation() {
  return this.appModel.orientation;
};

commands.setOrientation = async function setOrientation(o) {
  if (!_lodash.default.includes(['LANDSCAPE', 'PORTRAIT'], o)) {
    throw new _appiumBaseDriver.errors.UnknownError('Orientation must be LANDSCAPE or PORTRAIT');
  }

  this.appModel.orientation = o;
};

commands.getScreenshot = async function getScreenshot() {
  return this.appModel.getScreenshot();
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
