// appium-geckodriver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _logger = _interopRequireDefault(require("./logger"));

var _gecko = _interopRequireDefault(require("./gecko"));

var _desiredCaps = require("./desired-caps");

var _index = _interopRequireDefault(require("./commands/index"));

var _utils = require("./utils");

const NO_PROXY = [['GET', new RegExp('^/session/[^/]+/appium')], ['POST', new RegExp('^/session/[^/]+/appium')], ['POST', new RegExp('^/session/[^/]+/element/[^/]+/elements?$')], ['POST', new RegExp('^/session/[^/]+/elements?$')]];

class GeckoDriver extends _appiumBaseDriver.BaseDriver {
  constructor(opts = {}) {
    super(opts);
    this.desiredCapConstraints = _desiredCaps.desiredCapConstraints;
    this.locatorStrategies = ['xpath', 'tag name', 'link text', 'partial link text', 'css selector', 'id', 'name'];
    this.resetState();

    for (const [cmd, fn] of _lodash.default.toPairs(_index.default)) {
      GeckoDriver.prototype[cmd] = fn;
    }
  }

  resetState() {
    this.gecko = null;
    this.proxyReqRes = null;
    this.isProxyActive = false;
  }

  proxyActive() {
    return this.isProxyActive;
  }

  getProxyAvoidList() {
    return NO_PROXY;
  }

  canProxy() {
    return true;
  }

  async createSession(...args) {
    const [sessionId, caps] = await super.createSession(...args);
    this.gecko = new _gecko.default(caps);

    try {
      await this.gecko.start((0, _utils.formatCapsForServer)(caps));
    } catch (e) {
      await this.deleteSession();
      throw e;
    }

    this.proxyReqRes = this.gecko.proxy.proxyReqRes.bind(this.gecko.proxy);
    this.isProxyActive = true;
    return [sessionId, caps];
  }

  async deleteSession() {
    _logger.default.info('Ending Gecko Driver session');

    await this.gecko.stop();
    this.resetState();
    await super.deleteSession();
  }

}

var _default = GeckoDriver;
exports.default = _default;require('source-map-support').install();
