// appium/build/lib/argsparse-actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StoreDeprecatedTrueAction = exports.StoreDeprecatedDefaultCapabilityTrueAction = exports.StoreDeprecatedDefaultCapabilityAction = exports.StoreDeprecatedAction = exports.DEFAULT_CAPS_ARG = void 0;

require("source-map-support/register");

var _argparse = require("argparse");

const DEFAULT_CAPS_ARG = '--default-capabilities';
exports.DEFAULT_CAPS_ARG = DEFAULT_CAPS_ARG;

class StoreDeprecatedAction extends _argparse.Action {
  constructor(options = {}) {
    const opts = Object.assign({}, options);
    let helpPrefix = '[DEPRECATED]';

    if (opts.deprecated_for) {
      helpPrefix = `[DEPRECATED, use ${opts.deprecated_for} instead]`;
      delete opts.deprecated_for;
    }

    if (opts.help) {
      opts.help = `${helpPrefix} - ${opts.help}`;
    } else {
      opts.help = helpPrefix;
    }

    super(opts);
  }

  call(parser, namespace, values) {
    namespace[this.dest] = values;
  }

}

exports.StoreDeprecatedAction = StoreDeprecatedAction;

class StoreDeprecatedTrueAction extends StoreDeprecatedAction {
  constructor(options = {}) {
    super(Object.assign({}, options, {
      const: true,
      nargs: 0
    }));
  }

  call(parser, namespace) {
    namespace[this.dest] = this.const;
  }

}

exports.StoreDeprecatedTrueAction = StoreDeprecatedTrueAction;

class StoreDeprecatedDefaultCapabilityAction extends StoreDeprecatedAction {
  constructor(options = {}) {
    super(Object.assign({}, options, {
      deprecated_for: DEFAULT_CAPS_ARG
    }));
  }

  _writeDefaultCap(namespace, value) {
    namespace[this.dest] = value;

    if (value === this.default) {
      return;
    }

    if (!namespace.defaultCapabilities) {
      namespace.defaultCapabilities = {};
    }

    namespace.defaultCapabilities[this.dest] = value;
  }

  call(parser, namespace, values) {
    this._writeDefaultCap(namespace, values);
  }

}

exports.StoreDeprecatedDefaultCapabilityAction = StoreDeprecatedDefaultCapabilityAction;

class StoreDeprecatedDefaultCapabilityTrueAction extends StoreDeprecatedDefaultCapabilityAction {
  constructor(options = {}) {
    super(Object.assign({}, options, {
      const: true,
      nargs: 0
    }));
  }

  call(parser, namespace) {
    this._writeDefaultCap(namespace, this.const);
  }

}

exports.StoreDeprecatedDefaultCapabilityTrueAction = StoreDeprecatedDefaultCapabilityTrueAction;require('source-map-support').install();
