// appium-fake-driver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FakeDriver = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _fakeApp = require("./fake-app");

var _commands = _interopRequireDefault(require("./commands"));

class FakeDriver extends _appiumBaseDriver.BaseDriver {
  constructor() {
    super();
    this.appModel = null;
    this.curContext = 'NATIVE_APP';
    this.elMap = {};
    this.focusedElId = null;
    this.maxElId = 0;
    this.caps = {};
    this.desiredCapConstraints = {
      app: {
        presence: true,
        isString: true
      }
    };
  }

  async createSession(desiredCaps, requiredCaps, capabilities, otherSessionData = []) {
    for (let d of otherSessionData) {
      if (d.isUnique) {
        throw new _appiumBaseDriver.errors.SessionNotCreatedError('Cannot start session; another ' + 'unique session is in progress that requires all resources');
      }
    }

    let [sessionId, caps] = await super.createSession(desiredCaps, requiredCaps, capabilities, otherSessionData);
    this.appModel = new _fakeApp.FakeApp();

    if (_lodash.default.isArray(caps) === true && caps.length === 1) {
      caps = caps[0];
    }

    this.caps = caps;
    await this.appModel.loadApp(caps.app);
    return [sessionId, caps];
  }

  get driverData() {
    return {
      isUnique: !!this.caps.uniqueApp
    };
  }

}

exports.FakeDriver = FakeDriver;

for (let [cmd, fn] of _lodash.default.toPairs(_commands.default)) {
  FakeDriver.prototype[cmd] = fn;
}require('source-map-support').install();
