// appium-fake-driver/build/lib/fake-element.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FakeElement = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _xmldom = _interopRequireDefault(require("xmldom"));

class FakeElement {
  constructor(xmlNode, app) {
    this.app = app;
    this.node = xmlNode;
    this.nodeAttrs = {};
    this.type = this.node.tagName;
    this.attrs = {};
    this.css = {};

    for (let {
      name,
      value
    } of _lodash.default.values(this.node.attributes)) {
      this.nodeAttrs[name] = value;
    }

    this.parseCss();
  }

  parseCss() {
    if (this.nodeAttrs.style) {
      let segments = this.nodeAttrs.style.split(';');

      for (let s of segments) {
        let [prop, val] = s.split(':');
        prop = prop.trim();
        val = val.trim();
        this.css[prop] = val;
      }
    }
  }

  get tagName() {
    return this.node.tagName;
  }

  setAttr(k, v) {
    this.attrs[k] = v;
  }

  getAttr(k) {
    return this.attrs[k] || '';
  }

  isVisible() {
    return this.nodeAttrs.visible !== 'false';
  }

  isEnabled() {
    return this.nodeAttrs.enabled !== 'false';
  }

  isSelected() {
    return this.nodeAttrs.selected === 'true';
  }

  getLocation() {
    return {
      x: parseFloat(this.nodeAttrs.left || 0),
      y: parseFloat(this.nodeAttrs.top || 0)
    };
  }

  getSize() {
    return {
      width: parseFloat(this.nodeAttrs.width || 0),
      height: parseFloat(this.nodeAttrs.height || 0)
    };
  }

  click() {
    let curClicks = this.getAttr('clicks') || 0;
    this.setAttr('clicks', curClicks + 1);
    let alertId = this.nodeAttrs.showAlert;

    if (alertId) {
      this.app.showAlert(alertId);
    }
  }

  equals(other) {
    return this.node === other.node;
  }

  hasPrompt() {
    return this.nodeAttrs.hasPrompt === 'true';
  }

  getCss(prop) {
    if (_lodash.default.has(this.css, prop)) {
      return this.css[prop];
    }

    return null;
  }

  get xmlFragment() {
    let frag = new _xmldom.default.XMLSerializer().serializeToString(this.node);
    return new _xmldom.default.DOMParser().parseFromString(frag, 'application/xml');
  }

}

exports.FakeElement = FakeElement;require('source-map-support').install();
